joomla/joomla-cms
# [4.0][webservices] allow public GET with rate limit control

Pull Request for Issue #27569  point 3).

### Summary of Changes

-  endpoint fine grained control
-  allow public GET for an endpoint
-  endpoint rate limit control
- CLI command to reset limit quota `php cli/joomla.php webservices:reset`

enabled for:
 

- `plg_webservices_content`
- `plg_webservices_contact`
- `plg_webservices_modules`
- `plg_webservices_users`

the others webservices plugin works as before (ie no public GET)


### Testing Instructions

- apply PR or download the Prebuilt package
- generate a new Joomla API TOKEN
- go to plugins - Web Services - Content/Contact/Modules/Users
- check the new config parameters

![image](https://user-images.githubusercontent.com/181681/104038795-d2b6a080-51d5-11eb-8f9e-1792a2fd446a.png)


### Endpoint Allowed Verbs Test
- select all verbs and check that each one works as before
- check that only the allowed verbs works
-- enable POST only
-- and make a GET {{base_path}}/api/index.php/v1/content/article 
-- you'll get a **404 - Resource not found**

### Public GET Test
- enable Allow Public GET
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **without** the Joomla API TOKEN (noAuth)
-- you'll get a **200 response**
-- check that only public content is returned

- enable Allow Public GET
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **with** the Joomla API TOKEN 
-- you'll get a **200 response**
-- check that all content is returned (punblishe/unpublished)

- disable Allow Public GET
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **without** the Joomla API TOKEN (noAuth)
-- you'll get a **401 response**

- disable Allow Public GET
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **with** the Joomla API TOKEN 
-- you'll get a **200 response**

### Rate Limit Test
- enable Allow Public GET
- set Rate limit to 5
-- and make 6 GET {{base_path}}/api/index.php/v1/content/article  **without** the Joomla API TOKEN (noAuth)
-- you'll get a **429 response**
- click on Reset Limit
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **without** the Joomla API TOKEN (noAuth)
-- you'll get a **200 response**

### Actual result BEFORE applying this Pull Request

n/a

### Expected result AFTER applying this Pull Request

you can allow public GET
you can choose what verbs can be requested for each plugin

### Documentation Changes Required
yes
